### PR TITLE
feat(processor): forward events to new events-schema jobsDB

### PR DESCRIPTION
**Pull request by Sidddddarth**
# Description

Processor writes jobs to a new jobsdb - eventSchemasDB as well.

## Notion Ticket

[event schemas jobs db](https://www.notion.so/rudderstacks/new-jobsdb-for-event-schema-4f3d9595f0164a4e98f46b4ca58f8b8f?pvs=4)

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

File: processor/manager_test.go [hunk 1]
```diff
@@ -201,7 +201,7 @@ func TestProcessorManager(t *testing.T) {
 			"batch_rt": &jobsdb.MultiTenantLegacy{HandleT: brtDB},
 		},
 	}
-	processor := New(ctx, &clearDb, gwDB, rtDB, brtDB, errDB, mtStat, &reporting.NOOP{}, transientsource.NewEmptyService(), fileuploader.NewDefaultProvider(), mockRsourcesService, destinationdebugger.NewNoOpService(), transformationdebugger.NewNoOpService(),
+	processor := New(ctx, &clearDb, gwDB, rtDB, brtDB, errDB, nil, mtStat, &reporting.NOOP{}, transientsource.NewEmptyService(), fileuploader.NewDefaultProvider(), mockRsourcesService, destinationdebugger.NewNoOpService(), transformationdebugger.NewNoOpService(),
```
Comment (atzoum) [inline processor/manager_test.go:204]: let's use an actual db in this test

File: processor/processor.go [hunk 2]
```diff
@@ -1350,18 +1367,57 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob, parsedE
 					source,
 				)
 
+				payloadFunc := func() json.RawMessage {
```
Comment (atzoum) [inline processor/processor.go:1370]: we are effectively calling this twice. We can memoize it by assigning the result to a variable and returning this variable during subsequent calls

File: processor/processor.go [hunk 3]
```diff
@@ -1350,18 +1367,57 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob, parsedE
...
+					}
+					return payloadBytes
+				}
+				if proc.config.eventSchemaV2Enabled && // schemas enabled
+					// source has schemas enabled or if we override schemas for all sources
+					(source.EventSchemasEnabled || proc.config.eventSchemaV2AllSources) &&
+					// TODO: could use source.SourceDefinition.Category instead?
+					commonMetadataFromSingularEvent.SourceJobRunID == "" {
+					payload := payloadFunc()
+					if payload != nil {
```
Comment (atzoum) [inline processor/processor.go:1386]:
```suggestion
					if payload := payloadFunc(); payload != nil {
```

File: processor/processor.go (hunk 2 above)
Comment (atzoum) [inline processor/processor.go:1370]: first call assumes `nil` as the default and the second one assumes `{}`
```suggestion
				payloadFunc := func(default json.RawMessage) json.RawMessage {
```

File: processor/processor.go [hunk 4]
```diff
@@ -1410,6 +1466,27 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob, parsedE
 		statusList = append(statusList, &newStatus)
 	}
 
+	if len(eventSchemaJobs) > 0 {
+		proc.logger.Debug("[Processor] Total jobs written to event_schema: ", len(eventSchemaJobs))
```
Comment (atzoum) [inline processor/processor.go:1470]: this debug statement is misleading since nothing has been written here

File: processor/processor.go [hunk 5]
```diff
@@ -1350,18 +1367,68 @@ func (proc *Handle) processJobsForDest(partition string, subJobs subJob, parsedE
...
+								return payloadBytes
+							}()
+							called = true
+						}
+						if actualPayload == nil {
+							return defaultPayload
+						}
+						return actualPayload
+					}
+				}()
```
Comment (atzoum) [inline processor/processor.go:1394]: This appears to have too much code. Can we do the following?

1. We create a reusable function in rudder-go-kit
```go
package ro

func Memoize[R any](f func() R) func() R {
	var result R
	var called bool
	return func() R {
		if called {
			return result
		}
		result = f()
		return result
	}
}
```
2. We use it like so
```go
payloadFunc := ro.Memoize(func() json.RawMessage {
	if proc.transientSources.Apply(source.ID) {
		return nil
	}
	payloadBytes, err := jsonfast.Marshal(singularEvent)
	if err != nil {
		return nil
	}
	return payloadBytes
})

...

if payload := payloadFunc(); payload != nil {

...


func() json.RawMessage {
	if p := payloadFunc(); p != nil {
		return p
	}
	return []byte("{}")
},
```

Comment (Sidddddarth) [reply processor/processor.go:1394]: [raised a PR here](https://github.com/rudderlabs/rudder-go-kit/pull/30)
Will change here after that goes in.

Comment (atzoum) [reply processor/processor.go:1394]: https://github.com/rudderlabs/rudder-go-kit/releases/tag/v0.13.0

**Review by atzoum — APPROVED** [review]

**Review by cisse21 — APPROVED** [review]

**Pull request merged**